candid/src/parser/value.rs

```diff
@@ -319,6 +319,78 @@ impl IDLValue {
     }
 }
 
+pub mod pretty {
+    use super::*;
+    use crate::pretty::*;
+
+    use ::pretty::RcDoc;
+
+    pub fn pp_label(id: &Label) -> RcDoc {
+        match id {
+            Label::Named(id) => str("\"")
+                .append(format!("{}", id.escape_debug()))
+                .append("\""),
+            Label::Id(n) | Label::Unnamed(n) => RcDoc::as_string(n),
+        }
+    }
+
+    pub fn pp_idl_field(field: &IDLField) -> RcDoc {
+        pp_label(&field.id)
+            .append(kwd(" ="))
+            .append(pp_value(&field.val))
+    }
+
+    pub fn pp_idl_fields(fields: &[IDLField]) -> RcDoc {
+        concat(fields.iter().map(|f| pp_idl_field(f)), ";")
+    }
+
+    pub fn pp_variant_field(field: &IDLField) -> RcDoc {
+        pp_label(&field.id)
+            .append(kwd(" ="))
+            .append(pp_value(&field.val))
+    }
+
+    pub fn pp_value(v: &IDLValue) -> RcDoc {
+        use super::IDLValue::*;
+        match &*v {
+            Bool(true) => str("true"),
+            Bool(false) => str("false"),
+            Null => str("null"),
+            None => str("none"),
+            Int(i) => RcDoc::as_string(i),
+            Nat(n) => RcDoc::as_string(n),
+            Nat8(n) => RcDoc::as_string(n),
+            Nat16(n) => RcDoc::as_string(n),
+            Nat32(n) => RcDoc::as_string(n),
+            Nat64(n) => RcDoc::as_string(n),
+            Int8(i) => RcDoc::as_string(i),
+            Int16(i) => RcDoc::as_string(i),
+            Int32(i) => RcDoc::as_string(i),
+            Int64(i) => RcDoc::as_string(i),
+            Number(t) => str(t),
+            Text(t) => RcDoc::as_string(format!("{:?}", t)), // to do -- enough quoting here?
+            Opt(v) => kwd("opt").append(enclose_space("{", pp_value(v), "}")),
+            Vec(vs) => {
+                let mut body = RcDoc::nil();
+                let mut is_first = true;
+                for v in vs.iter() {
+                    if is_first {
+                        is_first = false;
+                    } else {
+                        body = body.append(RcDoc::space())
+                    }
+                    body = body.append(pp_value(v).append(RcDoc::text(";")))
```

Comment:
I'm a bit confused here. Why is `concat` or `RcDoc::intersperse` not enough here?

Reply:
I wrote this code by hand because the higher order options seemed more confusing, not simpler.

BTW, without an example, `concat` is confusing, and the comment `If it is displayed in a single line, omit the last separator.` is also confusing too.  This behavior is about the length of the sequence, not the 2D layout on the screen (space or line), right?

Reply:
It's about the 2D layout on screen. Let's say to render `{1;2;3;}`, if it's rendered in a single line, you get `{1;2;3}`, if not, you get
```
{
  1;
  2;
  3;
}
```